### PR TITLE
Add speculative decoding params to lm_bench

### PR DESCRIPTION
**Pull request by sbalandi**
Task: [CVS-155520](https://jira.devtools.intel.com/browse/CVS-155520)

### CODE REVIEW

File: tools/llm_bench/benchmark.py [hunk 1]
```diff
@@ -140,6 +140,12 @@ def get_argprser():
     parser.add_argument('--lora_alphas', nargs='*', help='Alphas params for LoRA adapters.', required=False, default=[])
     parser.add_argument("--use_cb", action="store_true", help="Use Continuous Batching inference mode")
     parser.add_argument("--cb_config", required=False, default=None, help="Path to file with Continuous Batching Scheduler settings or dict")
+    parser.add_argument("--draft_model", required=False, default=None, 
+                        help="Path to draft model folder including IR files for Speculative decoding generation.")
+    parser.add_argument("--draft_device", required=False, default='cpu', help="Inference device for Speculative decoding generation.")
+    parser.add_argument("--num_assistant_tokens", required=False, default=5, help="Config option num_assistant_tokens for Speculative decoding")
+    parser.add_argument("--assistant_confidence_threshold", required=False, default=0,
```
Comment (eaidova) [inline tools/llm_bench/benchmark.py:147]: as I understand, the default behaviour is not set it at all if you want to use default constant number of candidates, not sure that 0 is follow to this logic

**Comment by eaidova** [comment]
@sbalandi could you please include test into GHA for speculative decognig case

File: tools/llm_bench/benchmark.py [hunk 2]
```diff
@@ -140,6 +140,12 @@ def get_argprser():
     parser.add_argument('--lora_alphas', nargs='*', help='Alphas params for LoRA adapters.', required=False, default=[])
     parser.add_argument("--use_cb", action="store_true", help="Use Continuous Batching inference mode")
     parser.add_argument("--cb_config", required=False, default=None, help="Path to file with Continuous Batching Scheduler settings or dict")
+    parser.add_argument("--draft_model", required=False, default=None, 
+                        help="Path to draft model folder including IR files for Speculative decoding generation.")
+    parser.add_argument("--draft_device", required=False, default='cpu', help="Inference device for Speculative decoding generation.")
+    parser.add_argument("--num_assistant_tokens", required=False, default=5, help="Config option num_assistant_tokens for Speculative decoding")
+    parser.add_argument("--assistant_confidence_threshold", required=False, default=0, 
+                        help="Config option assistant_confidence_threshold for Speculative decodin")
```
Comment (ilya-lavrenov) [inline tools/llm_bench/benchmark.py:148]:
```suggestion
                        help="Config option assistant_confidence_threshold for Speculative decoding")
```

File: tools/llm_bench/llm_bench_utils/model_utils.py [hunk 3]
```diff
@@ -135,12 +135,18 @@ def analyze_args(args):
...
-    if args.use_cb and not args.genai:
+    if (args.use_cb or args.draft_model) and not args.genai:
         raise RuntimeError("Continuous batching mode supported only via OpenVINO GenAI")
     cb_config = None
     if args.cb_config:
         cb_config = get_config(args.cb_config)
     model_args["cb_config"] = cb_config
+    model_args['draft_model'] = args.draft_model
+    model_args['draft_device'] = args.draft_device
+    if (args.num_assistant_tokens > 0 and args.assistant_confidence_threshold > 0):
```
Comment (iefode) [inline tools/llm_bench/llm_bench_utils/model_utils.py:146]: Check please that one arg is 0 and other is not 0 in SpecDecoding case

File: tools/llm_bench/benchmark.py [hunk 4]
```diff
@@ -140,6 +140,14 @@ def get_argprser():
     parser.add_argument('--lora_alphas', nargs='*', help='Alphas params for LoRA adapters.', required=False, default=[])
     parser.add_argument("--use_cb", action="store_true", help="Use Continuous Batching inference mode")
     parser.add_argument("--cb_config", required=False, default=None, help="Path to file with Continuous Batching Scheduler settings or dict")
+    parser.add_argument("--draft_model", required=False, default=None,
+                        help="Path to draft model folder including IR files for Speculative decoding generation")
+    parser.add_argument("--draft_device", required=False, default=None, help="Inference device for Speculative decoding generation")
```
Comment (eaidova) [inline tools/llm_bench/benchmark.py:145]:
```suggestion
    parser.add_argument("--draft_device", required=False, default=None, help="Inference device for Speculative decoding draft model")
```

File: tools/llm_bench/benchmark.py [hunk 5]
```diff
@@ -140,6 +140,14 @@ def get_argprser():
...
+    parser.add_argument("--draft_cb_config", required=False, default=None,
+                        help="Path to file with Continuous Batching Scheduler settings or dict for Speculative decoding generation")
```
Comment (eaidova) [inline tools/llm_bench/benchmark.py:147]:
```suggestion
                        help="Path to file with Continuous Batching Scheduler settings or dict for Speculative decoding draft model")
```

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 6]
```diff
@@ -227,6 +228,13 @@ def create_genai_text_gen_model(model_path, device, ov_config, **kwargs):
                 setattr(scheduler_config, param, value)
         ov_config["scheduler_config"] = scheduler_config
 
+    if draft_model_path:
+        if not Path(draft_model_path).exists():
+            raise RuntimeError(f'==Failure ==: draft model by path:{draft_model_path} is not exists')
+        log.info("Speculative Decoding is activated")
+        draft_device = kwargs['draft_device'].upper() if kwargs.get('draft_device', None) else device.upper()
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:235]:
```suggestion
        draft_device = kwargs.get('draft_device',  device).upper()
```

**Comment by eaidova** [comment]
@sbalandi looks like the selected for test models are too large. Maybe we can use something less compute extensive? e.g. tinyllama with fp16 and int4/int8 precision as draft. Also you can use preconverted models from here https://huggingface.co/collections/OpenVINO/llm-6687aaa2abca3bbcec71a9bd

File: tools/llm_bench/benchmark.py [hunk 7]
```diff
@@ -140,6 +140,14 @@ def get_argprser():
     parser.add_argument('--lora_alphas', nargs='*', help='Alphas params for LoRA adapters.', required=False, default=[])
     parser.add_argument("--use_cb", action="store_true", help="Use Continuous Batching inference mode")
     parser.add_argument("--cb_config", required=False, default=None, help="Path to file with Continuous Batching Scheduler settings or dict")
+    parser.add_argument("--draft_model", required=False, default=None,
+                        help="Path to draft model folder including IR files for Speculative decoding generation")
+    parser.add_argument("--draft_device", required=False, default=None, help="Inference device for Speculative decoding of draft model")
+    parser.add_argument("--draft_cb_config", required=False, default=None,
```
Comment (eaidova) [inline tools/llm_bench/benchmark.py:146]: I do not see any usage of this cb config

Comment (sbalandi) [reply tools/llm_bench/benchmark.py:146]: added

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 8]
```diff
@@ -218,22 +233,15 @@ def create_genai_text_gen_model(model_path, device, ov_config, **kwargs):
...
 
     if draft_model_path:
         if not Path(draft_model_path).exists():
             raise RuntimeError(f'==Failure ==: draft model by path:{draft_model_path} is not exists')
         log.info("Speculative Decoding is activated")
-
-        ov_config['draft_model'] = openvino_genai.draft_model(draft_model_path, kwargs['draft_device'].upper())
+        draft_device = kwargs.get('draft_device', None) or device
+        ov_config['draft_model'] = openvino_genai.draft_model(draft_model_path, draft_device.upper(),
+                                                              scheduler_config=get_scheduler_config_genai("draft_cb_config", **kwargs))
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:244]: if draft model cb config is not provided, then it should be not set or used the same like for target model

Comment (eaidova) [reply tools/llm_bench/llm_bench_utils/ov_utils.py:244]: also you need to add parcing for this config file similar to original cb config https://github.com/openvinotoolkit/openvino.genai/blob/master/tools/llm_bench/llm_bench_utils/model_utils.py#L142

Comment (sbalandi) [reply tools/llm_bench/llm_bench_utils/ov_utils.py]: updated

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 9]
```diff
@@ -214,18 +229,19 @@ def create_genai_text_gen_model(model_path, device, ov_config, **kwargs):
...
+        ov_config["scheduler_config"] = get_scheduler_config_genai("cb_config", **kwargs)
 
-            for param, value in scheduler_params.items():
-                setattr(scheduler_config, param, value)
-        ov_config["scheduler_config"] = scheduler_config
+    if draft_model_path:
+        if not Path(draft_model_path).exists():
+            raise RuntimeError(f'==Failure ==: draft model by path:{draft_model_path} is not exists')
+        log.info("Speculative Decoding is activated")
+        draft_device = kwargs.get('draft_device', None) or device
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:242]:
```suggestion
        draft_device = kwargs.get('draft_device', device)
```

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 10]
```diff
@@ -204,6 +204,21 @@ def create_text_gen_model(model_path, device, **kwargs):
...
+    default_cb_config = {"cache_size": 1}
+    scheduler_config = openvino_genai.SchedulerConfig()
+    scheduler_params = kwargs.get(user_config) or default_cb_config
+    if scheduler_params:
+        log.info(f"Scheduler parameters for {user_config}:\n{scheduler_params}")
+
+        for param, value in scheduler_params.items():
+            setattr(scheduler_config, param, value)
+
+    return scheduler_config
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:219]:
```suggestion
def get_scheduler_config_genai(user_config, config_name=""CB config"):
    import openvino_genai

    default_cb_config = {"cache_size": 1}
    scheduler_config = openvino_genai.SchedulerConfig()
    scheduler_params = user_config or default_cb_config
    if scheduler_params:
        log.info(f"Scheduler parameters for {config_name}:\n{scheduler_params}")

        for param, value in scheduler_params.items():
            setattr(scheduler_config, param, value)

    return scheduler_config
```

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 11]
```diff
@@ -214,18 +229,19 @@ def create_genai_text_gen_model(model_path, device, ov_config, **kwargs):
...
     cb = kwargs.get("use_cb", False)
-    if cb:
+    if cb or draft_model_path:
         log.info("Continuous Batching mode activated")
-        default_cb_config = {"cache_size": 1}
-        scheduler_config = openvino_genai.SchedulerConfig()
-        scheduler_params = kwargs.get("cb_config") or default_cb_config
-        if scheduler_params:
-            log.info(f"Scheduler parameters:\n{scheduler_params}")
+        ov_config["scheduler_config"] = get_scheduler_config_genai("cb_config", **kwargs)
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:236]:
```suggestion
        ov_config["scheduler_config"] = get_scheduler_config_genai(kwargs.get("cb_config"))
```

File: tools/llm_bench/llm_bench_utils/ov_utils.py [hunk 12]
```diff
@@ -214,18 +229,19 @@ def create_genai_text_gen_model(model_path, device, ov_config, **kwargs):
...
-            for param, value in scheduler_params.items():
-                setattr(scheduler_config, param, value)
-        ov_config["scheduler_config"] = scheduler_config
+    if draft_model_path:
+        if not Path(draft_model_path).exists():
+            raise RuntimeError(f'==Failure ==: draft model by path:{draft_model_path} is not exists')
+        log.info("Speculative Decoding is activated")
+        draft_device = kwargs.get('draft_device', None) or device
+        scheduler_config = {'scheduler_config': get_scheduler_config_genai("draft_cb_config", **kwargs)} if kwargs.get("draft_cb_config") else {}
+        ov_config['draft_model'] = openvino_genai.draft_model(draft_model_path, draft_device.upper(), **scheduler_config)
```
Comment (eaidova) [inline tools/llm_bench/llm_bench_utils/ov_utils.py:244]:
```suggestion
        draft_model_load_kwargs = {'scheduler_config': get_scheduler_config_genai(kwargs["draft_cb_config"]], "draft CB config")} if kwargs.get("draft_cb_config") is not None else {}
        ov_config['draft_model'] = openvino_genai.draft_model(draft_model_path, draft_device.upper(), **draft_model_load_kwargs)
```

File: tools/llm_bench/llm_bench_utils/ov_utils.py (hunk 9 above)
Comment (sbalandi) [reply tools/llm_bench/llm_bench_utils/ov_utils.py:242]: draft_device is presented in kwargs, but the value is None, so the value of draft_device after kwargs.get('draft_device', device) will be None

File: tools/llm_bench/llm_bench_utils/ov_utils.py (hunk 10 above)
Comment (sbalandi) [reply tools/llm_bench/llm_bench_utils/ov_utils.py]: changed

File: tools/llm_bench/llm_bench_utils/ov_utils.py (hunk 11 above)
Comment (sbalandi) [reply tools/llm_bench/llm_bench_utils/ov_utils.py]: changed

File: tools/llm_bench/llm_bench_utils/ov_utils.py (hunk 12 above)
Comment (sbalandi) [reply tools/llm_bench/llm_bench_utils/ov_utils.py]: changed

**Review by eaidova — APPROVED** [review]

**Pull request merged**